core/corehttp/gateway_handler_unixfs_dir.go

```diff
@@ -110,34 +111,43 @@ func (i *gatewayHandler) serveDirectory(ctx context.Context, w http.ResponseWrit
 
 	// storage for directory listing
 	var dirListing []directoryItem
-	dirit := dir.Entries()
-	for dirit.Next() {
-		size := "?"
-		if s, err := dirit.Node().Size(); err == nil {
-			// Size may not be defined/supported. Continue anyways.
-			size = humanize.Bytes(uint64(s))
-		}
 
-		resolved, err := i.api.ResolvePath(ctx, ipath.Join(resolvedPath, dirit.Name()))
-		if err != nil {
+	// Optimization 1:
+	// List children without fetching their root blocks (fast, but no size info)
+	results, err := i.api.Unixfs().Ls(ctx, resolvedPath, options.Unixfs.ResolveChildren(false))
+	if err != nil {
+		internalWebError(w, err)
+		return
+	}
+	for link := range results {
+		if link.Err != nil {
 			internalWebError(w, err)
 			return
 		}
-		hash := resolved.Cid().String()
-
-		// See comment above where originalUrlPath is declared.
+		hash := link.Cid.String()
 		di := directoryItem{
-			Size:      size,
-			Name:      dirit.Name(),
-			Path:      gopath.Join(originalUrlPath, dirit.Name()),
+			Size:      "", // no size because we did not fetch child nodes
+			Name:      link.Name,
+			Path:      gopath.Join(originalUrlPath, link.Name),
 			Hash:      hash,
 			ShortHash: shortHash(hash),
 		}
 		dirListing = append(dirListing, di)
```

Comment:
Append with known sizes, gc aint that smart *yet*, prealoc pls.

Reply:
@Jorropo is https://github.com/ipfs/go-ipfs/pull/8853/commits/13ba0f7d7ccb78b77b420465e012383df8fd416e what you had in mind?

Reply:
I was thinking only of `dirListing` but that better. :)

I should push a refactor only PR that updates thoses `prealloc` find.